File: tools/resilience/circuit_breaker.py
"""
Circuit Breaker for VANA

This module implements the circuit breaker pattern for external service dependencies.
It provides protection against cascading failures and allows for graceful degradation.
"""

import enum
import functools
import logging
import threading
import time
from collections.abc import Callable
from typing import Any, Optional, TypeVar

# Set up logging
logger = logging.getLogger(__name__)

# Type variable for function return type
T = TypeVar("T")


class CircuitState(enum.Enum):
    """Circuit breaker states."""

    CLOSED = "closed"  # Normal operation, requests are allowed
    OPEN = "open"  # Circuit is open, requests are not allowed
    HALF_OPEN = "half_open"  # Testing if the service is back online


class CircuitBreaker:
    """
    Circuit Breaker for external service dependencies.

    This class implements the circuit breaker pattern to protect against
    cascading failures when external services are unavailable.
    """

    def __init__(
        self,
        name: str,
        failure_threshold: int = 5,
        reset_timeout: float = 60.0,
        half_open_max_calls: int = 1,
    ):
        """
        Initialize the circuit breaker.

        Args:
            name: Name of the circuit breaker
            failure_threshold: Number of failures before opening the circuit
            reset_timeout: Time in seconds before attempting to close the circuit
            half_open_max_calls: Maximum number of calls allowed in half-open state
        """
        self.name = name
        self.failure_threshold = failure_threshold
        self.reset_timeout = reset_timeout
        self.half_open_max_calls = half_open_max_calls

        # Circuit state
        self.state = CircuitState.CLOSED
        self.failure_count = 0
        self.last_failure_time = 0
        self.half_open_calls = 0

        # Lock for thread safety
        self.lock = threading.RLock()

        # Listeners for state changes
        self.state_change_listeners = []

        logger.info(f"Circuit breaker '{name}' initialized")

    def add_state_change_listener(
        self, listener: Callable[[str, CircuitState, CircuitState], None]
    ) -> None:
        """
        Add a listener for state changes.

        Args:
            listener: Function to call when the state changes
        """
        self.state_change_listeners.append(listener)

    def _notify_state_change(
        self, old_state: CircuitState, new_state: CircuitState
    ) -> None:
        """
        Notify listeners of a state change.

        Args:
            old_state: Previous state
            new_state: New state
        """
        for listener in self.state_change_listeners:
            try:
                listener(self.name, old_state, new_state)
            except Exception as e:
                logger.error(
                    f"Error in circuit breaker state change listener: {str(e)}"
                )

    def allow_request(self) -> bool:
        """
        Check if a request is allowed.

        Returns:
            True if the request is allowed, False otherwise
        """
        with self.lock:
            if self.state == CircuitState.CLOSED:
                return True

            if self.state == CircuitState.OPEN:
                # Check if it's time to try again
                current_time = time.time()
                if current_time - self.last_failure_time >= self.reset_timeout:
                    old_state = self.state
                    self.state = CircuitState.HALF_OPEN
                    self.half_open_calls = 0
                    logger.info(
                        f"Circuit breaker '{self.name}' transitioning from OPEN to HALF_OPEN"
                    )
                    self._notify_state_change(old_state, self.state)
                    return True
                return False

            if self.state == CircuitState.HALF_OPEN:
                # Allow a limited number of calls in half-open state
                if self.half_open_calls < self.half_open_max_calls:
                    self.half_open_calls += 1
                    return True
                return False

            return False

    def on_success(self) -> None:
        """Record a successful call."""
        with self.lock:
            if self.state == CircuitState.HALF_OPEN:
                # If successful in half-open state, close the circuit
                old_state = self.state
                self.state = CircuitState.CLOSED
                self.failure_count = 0
                self.half_open_calls = 0
                logger.info(
                    f"Circuit breaker '{self.name}' transitioning from HALF_OPEN to CLOSED"
                )
                self._notify_state_change(old_state, self.state)
            elif self.state == CircuitState.CLOSED:
                # Reset failure count on success
                self.failure_count = 0

    def on_failure(self) -> None:
        """Record a failed call."""
        with self.lock:
            current_time = time.time()
            self.last_failure_time = current_time

            if self.state == CircuitState.HALF_OPEN:
                # If failed in half-open state, open the circuit again
                old_state = self.state
                self.state = CircuitState.OPEN
                logger.warning(
                    f"Circuit breaker '{self.name}' transitioning from HALF_OPEN to OPEN"
                )
                self._notify_state_change(old_state, self.state)
            elif self.state == CircuitState.CLOSED:
                # Increment failure count
                self.failure_count += 1

                # If failure threshold reached, open the circuit
                if self.failure_count >= self.failure_threshold:
                    old_state = self.state
                    self.state = CircuitState.OPEN
                    logger.warning(
                        f"Circuit breaker '{self.name}' transitioning from CLOSED to OPEN after {self.failure_count} failures"
                    )
                    self._notify_state_change(old_state, self.state)

    def get_state(self) -> dict[str, Any]:
        """
        Get the current state of the circuit breaker.

        Returns:
            Dictionary with circuit breaker state information
        """
        with self.lock:
            return {
                "name": self.name,
                "state": self.state.value,
                "failure_count": self.failure_count,
                "failure_threshold": self.failure_threshold,
                "last_failure_time": self.last_failure_time,
                "reset_timeout": self.reset_timeout,
                "half_open_calls": self.half_open_calls,
                "half_open_max_calls": self.half_open_max_calls,
            }

    def reset(self) -> None:
        """Reset the circuit breaker to closed state."""
        with self.lock:
            old_state = self.state
            self.state = CircuitState.CLOSED
            self.failure_count = 0
            self.half_open_calls = 0
            logger.info(f"Circuit breaker '{self.name}' manually reset to CLOSED")
            self._notify_state_change(old_state, self.state)

    def __call__(self, func: Callable[..., T]) -> Callable[..., T]:
        """
        Decorator for protecting a function with a circuit breaker.

        Args:
            func: Function to protect

        Returns:
            Protected function
        """

        @functools.wraps(func)
        def wrapper(*args: Any, **kwargs: Any) -> T:
            if not self.allow_request():
                logger.warning(
                    f"Circuit breaker '{self.name}' is OPEN, request rejected"
                )
                raise CircuitBreakerOpenError(f"Circuit breaker '{self.name}' is open")

            try:
                result = func(*args, **kwargs)
                self.on_success()
                return result
            except Exception:
                self.on_failure()
                raise

        return wrapper


class CircuitBreakerOpenError(Exception):
    """Exception raised when a circuit breaker is open."""

    pass


class CircuitBreakerRegistry:
    """
    Registry for circuit breakers.

    This class provides a central registry for circuit breakers in the application.
    """

    def __init__(self):
        """Initialize the circuit breaker registry."""
        self.circuit_breakers = {}
        self.lock = threading.RLock()

        # Default state change listener
        self.add_state_change_listener(self._log_state_change)

        logger.info("Circuit breaker registry initialized")

    def get_or_create(
        self,
        name: str,
        failure_threshold: int = 5,
        reset_timeout: float = 60.0,
        half_open_max_calls: int = 1,
    ) -> CircuitBreaker:
        """
        Get or create a circuit breaker.

        Args:
            name: Name of the circuit breaker
            failure_threshold: Number of failures before opening the circuit
            reset_timeout: Time in seconds before attempting to close the circuit
            half_open_max_calls: Maximum number of calls allowed in half-open state

        Returns:
            Circuit breaker instance
        """
        with self.lock:
            if name not in self.circuit_breakers:
                circuit_breaker = CircuitBreaker(
                    name=name,
                    failure_threshold=failure_threshold,
                    reset_timeout=reset_timeout,
                    half_open_max_calls=half_open_max_calls,
                )

                # Add state change listeners
                for listener in self.state_change_listeners:
                    circuit_breaker.add_state_change_listener(listener)

                self.circuit_breakers[name] = circuit_breaker
                logger.info(f"Created circuit breaker '{name}'")

            return self.circuit_breakers[name]

    def get(self, name: str) -> Optional[CircuitBreaker]:
        """
        Get a circuit breaker by name.

        Args:
            name: Name of the circuit breaker

        Returns:
            Circuit breaker instance or None if not found
        """
        return self.circuit_breakers.get(name)

    def get_all(self) -> dict[str, CircuitBreaker]:
        """
        Get all circuit breakers.

        Returns:
            Dictionary of circuit breakers
        """
        return self.circuit_breakers.copy()

    def get_states(self) -> dict[str, dict[str, Any]]:
        """
        Get the states of all circuit breakers.

        Returns:
            Dictionary of circuit breaker states
        """
        return {name: cb.get_state() for name, cb in self.circuit_breakers.items()}

    def reset_all(self) -> None:
        """Reset all circuit breakers."""
        for circuit_breaker in self.circuit_breakers.values():
            circuit_breaker.reset()

    def reset(self, name: str) -> bool:
        """
        Reset a specific circuit breaker.

        Args:
            name: Name of the circuit breaker

        Returns:
            True if the circuit breaker was reset, False if not found
        """
        circuit_breaker = self.get(name)
        if circuit_breaker:
            circuit_breaker.reset()
            return True
        return False

    # State change listener management
    state_change_listeners = []

    @classmethod
    def add_state_change_listener(
        cls, listener: Callable[[str, CircuitState, CircuitState], None]
    ) -> None:
        """
        Add a listener for state changes in all circuit breakers.

        Args:
            listener: Function to call when a circuit breaker state changes
        """
        cls.state_change_listeners.append(listener)

    @classmethod
    def _log_state_change(
        cls, name: str, old_state: CircuitState, new_state: CircuitState
    ) -> None:
        """
        Log circuit breaker state changes.

        Args:
            name: Name of the circuit breaker
            old_state: Previous state
            new_state: New state
        """
        logger.info(
            f"Circuit breaker '{name}' state changed from {old_state.value} to {new_state.value}"
        )


# Create a global registry
registry = CircuitBreakerRegistry()


def circuit_breaker(
    name: str,
    failure_threshold: int = 5,
    reset_timeout: float = 60.0,
    half_open_max_calls: int = 1,
) -> Callable[[Callable[..., T]], Callable[..., T]]:
    """
    Decorator for protecting a function with a circuit breaker.

    Args:
        name: Name of the circuit breaker
        failure_threshold: Number of failures before opening the circuit
        reset_timeout: Time in seconds before attempting to close the circuit
        half_open_max_calls: Maximum number of calls allowed in half-open state

    Returns:
        Decorator function
    """

    def decorator(func: Callable[..., T]) -> Callable[..., T]:
        circuit_breaker = registry.get_or_create(
            name=name,
            failure_threshold=failure_threshold,
            reset_timeout=reset_timeout,
            half_open_max_calls=half_open_max_calls,
        )
        return circuit_breaker(func)

    return decorator

File: supabase/functions/_shared/circuit-breaker.ts
/**
 * Circuit Breaker Pattern Implementation
 *
 * Provides resilience for external API calls (e.g., Gemini, OpenRouter) by:
 * - Failing fast when a service is unhealthy (open state)
 * - Gradually testing recovery (half-open state)
 * - Normal operation when healthy (closed state)
 *
 * State Machine:
 * ```
 *   [CLOSED] ---(failure threshold exceeded)---> [OPEN]
 *      ^                                            |
 *      |                                            v
 *      +----------(success)---- [HALF-OPEN] <---(timeout)
 *                                    |
 *                                    +---(failure)---> [OPEN]
 * ```
 *
 * NOTE: In Deno Edge Functions, module-level state may reset on cold starts.
 * This provides "soft" resilience within a function's lifetime but won't
 * persist across isolate restarts. This is acceptable for our use case.
 *
 * @module circuit-breaker
 */

import { CircuitBreakerOpenError } from './errors.ts';

/**
 * Circuit breaker states
 */
export type CircuitState = 'closed' | 'open' | 'half-open';

/**
 * Configuration options for the circuit breaker
 */
export interface CircuitBreakerConfig {
  /** Number of consecutive failures before opening circuit (default: 5) */
  failureThreshold?: number;
  /** Time in ms to wait before transitioning from open to half-open (default: 30000) */
  resetTimeoutMs?: number;
  /** Name for logging purposes (default: 'CircuitBreaker') */
  name?: string;
}

/**
 * Internal state tracking for the circuit breaker
 */
interface CircuitBreakerState {
  /** Current state of the circuit */
  state: CircuitState;
  /** Count of consecutive failures */
  failureCount: number;
  /** Timestamp when circuit was opened */
  openedAt: number | null;
  /** Last successful call timestamp */
  lastSuccessAt: number | null;
  /** Last failure call timestamp */
  lastFailureAt: number | null;
  /** Total calls made */
  totalCalls: number;
  /** Total failures */
  totalFailures: number;
  /** Total successes */
  totalSuccesses: number;
}

/**
 * Circuit Breaker class for wrapping external API calls.
 *
 * @example
 * ```typescript
 * const circuitBreaker = new CircuitBreaker({
 *   failureThreshold: 5,
 *   resetTimeoutMs: 30000,
 *   name: 'GeminiAPI'
 * });
 *
 * // Normal usage
 * const result = await circuitBreaker.call(() => callGemini(messages));
 *
 * // With fallback
 * const result = await circuitBreaker.call(
 *   () => callGemini(messages),
 *   () => callFallbackModel(messages)
 * );
 * ```
 */
export class CircuitBreaker {
  private readonly config: Required<CircuitBreakerConfig>;
  private state: CircuitBreakerState;

  constructor(config: CircuitBreakerConfig = {}) {
    this.config = {
      failureThreshold: config.failureThreshold ?? 5,
      resetTimeoutMs: config.resetTimeoutMs ?? 30000,
      name: config.name ?? 'CircuitBreaker'
    };

    this.state = {
      state: 'closed',
      failureCount: 0,
      openedAt: null,
      lastSuccessAt: null,
      lastFailureAt: null,
      totalCalls: 0,
      totalFailures: 0,
      totalSuccesses: 0
    };

    console.log(
      `[${this.config.name}] Initialized with threshold=${this.config.failureThreshold}, ` +
      `resetTimeout=${this.config.resetTimeoutMs}ms`
    );
  }

  /**
   * Get the current circuit state.
   */
  getState(): CircuitState {
    // Check if we should transition from open to half-open
    if (this.state.state === 'open' && this.shouldTransitionToHalfOpen()) {
      this.transitionTo('half-open');
    }
    return this.state.state;
  }

  /**
   * Get the current failure count.
   */
  getFailureCount(): number {
    return this.state.failureCount;
  }

  /**
   * Get circuit breaker statistics.
   */
  getStats(): {
    state: CircuitState;
    failureCount: number;
    totalCalls: number;
    totalSuccesses: number;
    totalFailures: number;
    successRate: number;
  } {
    const successRate = this.state.totalCalls > 0
      ? this.state.totalSuccesses / this.state.totalCalls
      : 1;

    return {
      state: this.getState(),
      failureCount: this.state.failureCount,
      totalCalls: this.state.totalCalls,
      totalSuccesses: this.state.totalSuccesses,
      totalFailures: this.state.totalFailures,
      successRate: Math.round(successRate * 100) / 100
    };
  }

  /**
   * Execute a function through the circuit breaker.
   *
   * @param fn - The async function to execute
   * @param fallback - Optional fallback function to call when circuit is open
   * @returns The result of fn or fallback
   * @throws CircuitBreakerOpenError if circuit is open and no fallback provided
   */
  async call<T>(
    fn: () => Promise<T>,
    fallback?: () => Promise<T>
  ): Promise<T> {
    this.state.totalCalls++;

    // Check current state (also handles open->half-open transition)
    const currentState = this.getState();

    if (currentState === 'open') {
      return this.handleOpenState(fallback);
    }

    // Execute the function
    try {
      const result = await fn();
      this.onSuccess();
      return result;
    } catch (error) {
      this.onFailure(error);
      throw error;
    }
  }

  /**
   * Manually reset the circuit breaker to closed state.
   * Useful for testing or administrative resets.
   */
  reset(): void {
    const previousState = this.state.state;
    this.state.state = 'closed';
    this.state.failureCount = 0;
    this.state.openedAt = null;

    if (previousState !== 'closed') {
      console.log(
        `[${this.config.name}] MANUAL RESET: ${previousState} -> closed`
      );

      // Log structured event for monitoring
      this.logStateChange(previousState, 'closed', 'manual_reset');
    }
  }

  /**
   * Check if the circuit should transition from open to half-open.
   */
  private shouldTransitionToHalfOpen(): boolean {
    if (this.state.state !== 'open' || this.state.openedAt === null) {
      return false;
    }
    return Date.now() - this.state.openedAt >= this.config.resetTimeoutMs;
  }

  /**
   * Handle the case when circuit is in open state.
   */
  private async handleOpenState<T>(
    fallback?: () => Promise<T>
  ): Promise<T> {
    const resetAt = new Date(
      (this.state.openedAt ?? Date.now()) + this.config.resetTimeoutMs
    );

    console.log(
      `[${this.config.name}] Circuit OPEN - rejecting request. ` +
      `Resets at ${resetAt.toISOString()}`
    );

    // If fallback is provided, use it
    if (fallback) {
      const secondsUntilReset = Math.ceil((resetAt.getTime() - Date.now()) / 1000);

      console.log(
        `[${this.config.name}] WARNING: Using fallback function. ` +
        `Circuit is OPEN due to service degradation. ` +
        `Primary service will be retried in ${secondsUntilReset}s.`
      );

      // Log structured event for monitoring
      console.log(JSON.stringify({
        event: 'circuit_breaker_fallback_used',
        circuitName: this.config.name,
        secondsUntilReset,
        resetAt: resetAt.toISOString(),
        totalFailures: this.state.totalFailures,
        failureCount: this.state.failureCount,
        timestamp: new Date().toISOString()
      }));

      try {
        const result = await fallback();
        // Don't count fallback success as circuit recovery
        console.warn(
          `[${this.config.name}] Fallback succeeded but service is DEGRADED. ` +
          `Using secondary provider. Primary service will be retried in ${secondsUntilReset}s.`
        );
        return result;
      } catch (fallbackError) {
        console.error(
          `[${this.config.name}] CRITICAL: Fallback also failed. ` +
          `Both primary and fallback services unavailable. ` +
          `Circuit will reset in ${secondsUntilReset}s.`,
          fallbackError
        );
        throw fallbackError;
      }
    }

    // No fallback - throw error
    throw new CircuitBreakerOpenError(resetAt);
  }

  /**
   * Handle successful call completion.
   */
  private onSuccess(): void {
    this.state.totalSuccesses++;
    this.state.lastSuccessAt = Date.now();

    if (this.state.state === 'half-open') {
      // Success in half-open state -> close the circuit
      this.transitionTo('closed');
    }

    // Reset failure count on success (circuit is healthy)
    this.state.failureCount = 0;
  }

  /**
   * Handle call failure.
   */
  private onFailure(error: unknown): void {
    this.state.totalFailures++;
    this.state.failureCount++;
    this.state.lastFailureAt = Date.now();

    const errorMessage = error instanceof Error ? error.message : String(error);

    console.error(
      `[${this.config.name}] Call failed (count: ${this.state.failureCount}): ${errorMessage}`
    );

    if (this.state.state === 'half-open') {
      // Failure in half-open state -> reopen the circuit
      this.transitionTo('open');
    } else if (
      this.state.state === 'closed' &&
      this.state.failureCount >= this.config.failureThreshold
    ) {
      // Threshold exceeded -> open the circuit
      this.transitionTo('open');
    }
  }

  /**
   * Transition to a new state with logging.
   */
  private transitionTo(newState: CircuitState): void {
    const previousState = this.state.state;

    if (previousState === newState) {
      return; // No transition needed
    }

    this.state.state = newState;

    if (newState === 'open') {
      this.state.openedAt = Date.now();
    } else if (newState === 'closed') {
      this.state.failureCount = 0;
      this.state.openedAt = null;
    }

    console.log(
      `[${this.config.name}] State transition: ${previousState} -> ${newState}`
    );

    // Log structured event for monitoring
    this.logStateChange(previousState, newState, 'automatic');
  }

  /**
   * Log state change event as structured JSON for monitoring.
   */
  private logStateChange(
    from: CircuitState,
    to: CircuitState,
    trigger: 'automatic' | 'manual_reset'
  ): void {
    // Calculate success rate
    const successRate = this.state.totalCalls > 0
      ? this.state.totalSuccesses / this.state.totalCalls
      : 1;

    // Structured event for all state changes
    const event = {
      event: 'circuit_breaker_state_change',
      circuitName: this.config.name,
      from,
      to,
      trigger,
      failures: this.state.failureCount,
      totalCalls: this.state.totalCalls,
      totalFailures: this.state.totalFailures,
      totalSuccesses: this.state.totalSuccesses,
      successRate: Math.round(successRate * 100) / 100,
      timestamp: new Date().toISOString()
    };

    console.log(JSON.stringify(event));

    // ERROR-level logging when circuit opens (critical event)
    if (to === 'open') {
      const resetAt = new Date(Date.now() + this.config.resetTimeoutMs);
      console.error(
        `🚨 [${this.config.name}] CIRCUIT BREAKER OPENED - Service degraded! ` +
        `Failures: ${this.state.failureCount}/${this.config.failureThreshold}, ` +
        `Success rate: ${Math.round(successRate * 100)}%, ` +
        `Resets at: ${resetAt.toISOString()}`
      );

      // TODO: Integrate with monitoring service (Sentry/Datadog/CloudWatch)
      // Example: Sentry.captureMessage('Circuit breaker opened', { level: 'error', extra: event });
    }
  }
}

/**
 * Create a pre-configured circuit breaker for Gemini API calls.
 *
 * Default configuration:
 * - 5 failures to open (handles temporary blips)
 * - 30 second reset timeout (gives API time to recover)
 *
 * @returns Configured CircuitBreaker instance
 */
export function createGeminiCircuitBreaker(): CircuitBreaker {
  return new CircuitBreaker({
    failureThreshold: 5,
    resetTimeoutMs: 30000,
    name: 'GeminiAPI'
  });
}
